error cocurent logging

```
panic: runtime error: index out of range [1] with length 1

goroutine 1309 [running]:
github.com/rverst/zgelf.(*logBuffer).Copy(...)
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/buffer.go:32
github.com/rverst/zgelf.(*GelfWriter).Flush(0xc000190700, 0x0)
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/zgelf.go:128 +0x32d
github.com/rverst/zgelf.(*GelfWriter).bufferEvent(0xc000190700, 0xc00b442c60?)
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/zgelf.go:220 +0x15e
github.com/rverst/zgelf.(*GelfWriter).process(0xc000190700, 0xc00b442c30)
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/zgelf.go:208 +0x76a
github.com/rverst/zgelf.(*GelfWriter).worker.func1(0xc0007befb8?)
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/zgelf.go:160 +0x65
created by github.com/rverst/zgelf.(*GelfWriter).worker
        /home/*****/go/pkg/mod/github.com/rverst/zgelf@v0.0.0-20200728153303-fccb02ec5899/zgelf.go:158 +0x36
```
this patch may fix the error:

```diff
Index: buffer.go
<+>UTF-8
===================================================================
diff --git a/buffer.go b/buffer.go
--- a/buffer.go	(revision fccb02ec5899f90ca1583e7ff9ed49e5a10b1adf)
+++ b/buffer.go	(date 1655975317137)
@@ -2,11 +2,13 @@
 
 import (
 	"fmt"
+	"sync"
 )
 
 type logBuffer struct {
-	buffers   [][]byte
-	size      int
+	buffers [][]byte
+	size    int
+	mu      sync.RWMutex
 }
 
 func NewLogBuffer() *logBuffer {
@@ -17,16 +19,22 @@
 }
 
 func (b *logBuffer) Add(data []byte) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
 	b.size = b.size + len(data)
 	b.buffers = append(b.buffers, data)
 }
 
 func (b *logBuffer) Clear() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
 	b.size = 0
 	b.buffers = make([][]byte, 0)
 }
 
 func (b *logBuffer) Copy() *logBuffer {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
 	c := make([][]byte, len(b.buffers))
 	for i, x := range b.buffers {
 		c[i] = make([]byte, len(x))
@@ -43,6 +51,8 @@
 	if len(b.buffers) == 0 {
 		return nil, fmt.Errorf("no element left in buffer")
 	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
 	r := b.buffers[:1]
 	b.buffers = b.buffers[1:]
 	b.size -= len(r[0])
Index: zgelf.go
<+>UTF-8
===================================================================
diff --git a/zgelf.go b/zgelf.go
--- a/zgelf.go	(revision fccb02ec5899f90ca1583e7ff9ed49e5a10b1adf)
+++ b/zgelf.go	(date 1656333047977)
@@ -39,7 +39,6 @@
 	queue       chan map[string]interface{}
 	wgProcess   sync.WaitGroup
 	wgFlush     sync.WaitGroup
-	mu          sync.RWMutex
 	buffer      *logBuffer
 	ticker      *time.Ticker
 }
@@ -122,9 +121,6 @@
 		return
 	}
 
-	w.mu.Lock()
-	defer w.mu.Unlock()
-
 	c := w.buffer.Copy()
 	w.buffer.Clear()
 



```

--- a/buffer.go
+++ b/buffer.go
@@ -2,43 +2,56 @@ package zgelf
 
 import (
 	"fmt"
+	"sync"
 )
 
 type logBuffer struct {
-	buffers   [][]byte
-	size      int
+	buffers [][]byte
+	size    int
+	mu      sync.RWMutex
 }
 
 func NewLogBuffer() *logBuffer {
 	return &logBuffer{
-		buffers:   make([][]byte, 0),
-		size:      0,
+		buffers: make([][]byte, 0),
+		size:    0,
 	}
 }
 
 func (b *logBuffer) Add(data []byte) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
 	b.size = b.size + len(data)
 	b.buffers = append(b.buffers, data)
 }
 
 func (b *logBuffer) Clear() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
 	b.size = 0
 	b.buffers = make([][]byte, 0)
 }
 
 func (b *logBuffer) Copy() *logBuffer {
+	b.mu.RLock()
+	defer b.mu.Unlock()
+
 	c := make([][]byte, len(b.buffers))
 	for i, x := range b.buffers {
 		c[i] = make([]byte, len(x))
 		copy(c[i], x)
 	}
 	return &logBuffer{
-		buffers:   c,
-		size:      b.size,
+		buffers: c,
+		size:    b.size,
 	}
 }
 
 func (b *logBuffer) Pull() ([]byte, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
 
 	if len(b.buffers) == 0 {
 		return nil, fmt.Errorf("no element left in buffer")

--- a/buffer_test.go
+++ b/buffer_test.go
@@ -13,8 +13,6 @@ func TestNewLogBuffer(t *testing.T) {
 		{"create new logBuffer", &logBuffer{
 			buffers:   make([][]byte, 0),
 			size:      0,
-			compress:  false,
-			delimiter: make([]byte, 0),
 		}},
 	}
 	for _, tt := range tests {
@@ -57,9 +55,7 @@ func Test_logBuffer_Clear(t *testing.T) {
 	b2 := NewLogBuffer()
 
 	b1.Add([]byte(`{ "x":"Hello World" }`))
-	b1.compress = true
 	b2.Add([]byte(`{ "x":"Hello World" }`))
-	b2.SetDelimiter([]byte{'\n'})
 
 	tests := []struct {
 		name   string
@@ -74,14 +70,8 @@ func Test_logBuffer_Clear(t *testing.T) {
 			if tt.buffer.size > 0 {
 				t.Errorf("buffer.size error, want 0 got: %d", tt.buffer.size)
 			}
-			if tt.buffer.compress {
-				t.Errorf("buffer.compress error, want false got: %t", tt.buffer.compress)
-			}
 			if len(tt.buffer.buffers) > 0 {
 				t.Errorf("buffer.buffers not empty, got: %v", tt.buffer.buffers)
-			}
-			if len(tt.buffer.delimiter) > 0 {
-				t.Errorf("buffer.delimiter not empty, got: %v", tt.buffer.delimiter)
 			}
 		})
 	}

--- a/zgelf.go
+++ b/zgelf.go
@@ -5,12 +5,13 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"github.com/rs/zerolog"
 	"os"
 	"strconv"
 	"strings"
 	"sync"
 	"time"
+
+	"github.com/rs/zerolog"
 )
 
 const (
@@ -39,7 +40,6 @@ type GelfWriter struct {
 	queue       chan map[string]interface{}
 	wgProcess   sync.WaitGroup
 	wgFlush     sync.WaitGroup
-	mu          sync.RWMutex
 	buffer      *logBuffer
 	ticker      *time.Ticker
 }
@@ -70,7 +70,6 @@ func New(host, tmpLogPath string, trans transport) *GelfWriter {
 }
 
 func (w *GelfWriter) Write(p []byte) (n int, err error) {
-
 	var evt map[string]interface{}
 	d := json.NewDecoder(bytes.NewReader(p))
 	d.UseNumber()
@@ -113,7 +112,6 @@ func (w *GelfWriter) Close() {
 
 // Flush flushes the send buffer
 func (w *GelfWriter) Flush(block bool) {
-
 	if block {
 		w.wgProcess.Wait()
 		w.wgFlush.Wait() // wait for running, non blocking operations
@@ -121,9 +119,6 @@ func (w *GelfWriter) Flush(block bool) {
 	if w.buffer.Size() == 0 {
 		return
 	}
-
-	w.mu.Lock()
-	defer w.mu.Unlock()
 
 	c := w.buffer.Copy()
 	w.buffer.Clear()
@@ -152,7 +147,6 @@ func (w *GelfWriter) SetMaxBufferTime(bufferTime time.Duration) {
 }
 
 func (w *GelfWriter) worker() {
-
 	for data := range w.queue {
 		w.wgProcess.Add(1)
 		go func(evt map[string]interface{}) {
@@ -163,7 +157,6 @@ func (w *GelfWriter) worker() {
 }
 
 func (w *GelfWriter) process(evt map[string]interface{}) {
-
 	evn := make(map[string]interface{}, len(evt))
 	for k, v := range evt {
 		switch k {
@@ -245,7 +238,6 @@ func (w *GelfWriter) sendTemporaryLogs() {
 }
 
 func parseCaller(caller string) (file string, line int, err error) {
-
 	i := strings.LastIndex(caller, ":")
 	if i < 0 {
 		return "", 0, fmt.Errorf("cannot parse caller: %s", caller)

